### PR TITLE
Add sparse depth regularization

### PR DESCRIPTION
**Pull request by fwilliams**
Simple l1 loss on sparse depth in Gaussian splat training.

### CODE REVIEW

File: fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py [hunk 1]
```diff
@@ -1140,10 +1151,24 @@ def optimize(self, show_progress: bool = True, log_tag: str = "reconstruct") ->
...
+                )  # [B,]
 
                 # Progressively use higher spherical harmonic degree as we optimize
                 sh_degree_to_use = min(self._global_step // increase_sh_degree_every_step, self.config.sh_degree)
-                projected_gaussians = self.model.project_gaussians_for_images(
+                projection_function = (
+                    self.model.project_gaussians_for_images_and_depths
+                    if self.config.sparse_depth_reg > 0
+                    else self.model.project_gaussians_for_images
+                )
```
[the remark this answers is not in the document]
Comment (fwilliams) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py]: This method always exists and is part of the fvdb-core API

File: fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py [hunk 2]
```diff
@@ -1192,6 +1220,20 @@ def optimize(self, show_progress: bool = True, log_tag: str = "reconstruct") ->
                     )
                     loss = torch.lerp(l1loss, ssimloss, self.config.ssim_lambda)  # type: ignore
 
+                    if sparse_depth is not None and sparse_depth_uv is not None and median_depths is not None:
+                        depth = rgbd[..., 3]  # [1, H, W]
+                        depth_uv = depth[:, sparse_depth_uv[0, :, 1], sparse_depth_uv[0, :, 0]]  # [B, N]
+                        alpha_uv = alphas[:, sparse_depth_uv[0, :, 1], sparse_depth_uv[0, :, 0], 0]  # [B, N]
```
[the remark this answers is not in the document]
Comment (fwilliams) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py]: This code only works for batch size = 1. To make it work for batch size > 1, the dataset would need return a JaggedTensor of sparse pixel coordinates and depths (which requires modifying the collate function), and would either need a for loop to compute sampels, or to use the sparse rendering functions (which don't support duplicate pixels, a case that occurs in practice). Given that we almost always use BS=1 anyway, I'm going to add a check here and fail if the batch size exceeds 1. When we have duplicate pixel support for sparse rendering, we can update this code (CC @harrism)

File: fvdb_reality_capture/sfm_scene/sfm_scene.py [hunk 3]
```diff
@@ -460,6 +496,58 @@ def apply_transformation_matrix(self, transformation_matrix: np.ndarray) -> "Sfm
...
+            centroid = np.mean(origins, axis=0)
+            dists = np.linalg.norm(origins - centroid, axis=1)
+            return float(np.max(dists))
+        elif mode == SpatialScaleMode.SCENE_DIAGONAL_PERCENTILE:
+            percentiles = np.percentile(self.points, [5, 95], axis=1)
+            bbox_min, bbox_max = percentiles[:, 0], percentiles[:, 1]
+            scene_diag = np.linalg.norm(bbox_max - bbox_min)
+            return float(scene_diag)
+        else:
+            raise ValueError(f"Unknown spatial scale mode: {mode}")
```
[the remark this answers is not in the document]
Comment (harrism) [reply fvdb_reality_capture/sfm_scene/sfm_scene.py]: Agree.

File: fvdb_reality_capture/sfm_scene/sfm_scene.py [hunk 4]
```diff
@@ -12,6 +13,41 @@
 from .sfm_metadata import SfmCameraMetadata, SfmPosedImageMetadata
 
 
+class SpatialScaleMode(str, Enum):
```
Comment (harrism) [inline fvdb_reality_capture/sfm_scene/sfm_scene.py]: **💡 suggestion:** ‏I think the move of this class and the spatial_scale() utility method from the optimizer to the SfmScene deserves mention / reason in the PR description.

**Review by harrism — APPROVED** [review]
